source/Classes/1.DscResources/08.nxScript.ps1

```diff
@@ -0,0 +1,120 @@
+[DscResource()]
+class nxScript
+{
+    [DscProperty(Key)]
+    [System.String] $GetScript
+
+    [DscProperty(Key)]
+    [System.String] $TestScript
+
+    [DscProperty(Key)]
+    [System.String] $SetScript
+
+    [DscProperty()]
+    [Reason[]] $Reasons
+
+    [nxScript] Get()
+    {
+        $scriptBlock = [System.Management.Automation.ScriptBlock]::Create($this.GetScript)
+        $invokeScriptResult = $this.InvokeScript($scriptBlock)
+        $isValid = $this.TestGetScriptOutput($invokeScriptResult)
+
+        $currentState = [nxScript]::new()
+        $currentState.GetScript = $this.GetScript
+        $currentState.TestScript = $this.TestScript
+        $currentState.SetScript = $this.SetScript
+        if ($isValid)
+        {
+            $currentState.Reasons = $invokeScriptResult.Reasons
+        }
+
+        return $currentState
+    }
+
+    [bool] Test()
+    {
+        $scriptBlock = [System.Management.Automation.ScriptBlock]::Create($this.TestScript)
+        $invokeScriptResult = $this.InvokeScript($scriptBlock)
+
+        if ($invokeScriptResult -is [System.Management.Automation.ErrorRecord])
+        {
+            Write-Verbose -Message "TestScript returned an error."
+            return $false
+        }
+
+        if ($null -eq $invokeScriptResult -or -not ($invokeScriptResult -is [System.Boolean]))
+        {
+            Write-Verbose -Message "TestScript output is not a Boolean."
+            return $false
+        }
+
+        return $invokeScriptResult
+    }
+
+    [void] Set()
+    {
+        $scriptBlock = [System.Management.Automation.ScriptBlock]::Create($this.SetScript)
+        $this.InvokeScript($scriptBlock)
+    }
+
+    [System.Object] InvokeScript([System.Management.Automation.ScriptBlock] $ScriptBlock)
+    {
+        $scriptResult = $null
+        try
+        {
+            $scriptResult = & $ScriptBlock
+        }
+        catch
+        {
+            # Surfacing the error thrown by the execution of the script
+            $scriptResult = $_
+        }
+
+        return $scriptResult
+    }
+
+    [bool] TestGetScriptOutput([System.Object] $GetScriptOutput)
+    {
+        if ($GetScriptOutput -is [System.Management.Automation.ErrorRecord])
+        {
+            Write-Verbose -Message "GetScript returned an error."
+            return $false
+        }
+
+        if ($GetScriptOutput -isnot [System.Collections.Hashtable])
+        {
+            Write-Verbose -Message "GetScript output is not a hashtable."
```

Comment:
This should be worded more as an error? Right now it's just a Verbose message that doesn't sound like a problem.
Some thing like `GetScript must return a hashtable but it currently returns $($GetScriptOutput.GetType())` or `Error: GetScript output is not a hashtable.` or something along those lines.
I'm not sure if our engine currently prints the output of `Write-Error` but that would be the ideal option?

Reply:
Updated Get() to throw a detailed error if this returns False.